core/src/target_os_check.rs

```diff
@@ -0,0 +1,303 @@
+//! Optional checks for `#[cfg(target_os = "target")]
+use crate::parser::get_meta_items;
+use log::{debug, error, log_enabled, warn};
+use quote::ToTokens;
+use std::collections::VecDeque;
+use syn::{punctuated::Punctuated, Attribute, Expr, ExprLit, Lit, Meta, Token};
+
+#[derive(Copy, Clone, Default, Debug)]
+enum TargetScope {
+    #[default]
+    Accept,
+    Reject,
+}
+
+#[derive(Default)]
+struct TargetOsIterator {
+    meta: VecDeque<Meta>,
+    scope: TargetScope,
+}
+
+impl TargetOsIterator {
+    fn new(meta: Meta) -> Self {
+        Self {
+            meta: VecDeque::from([meta]),
+            ..Default::default()
+        }
+    }
+}
+
+impl Iterator for TargetOsIterator {
+    type Item = (TargetScope, String);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(meta) = self.meta.pop_front() {
+            if meta.path().is_ident("not") {
+                debug!("encountered not");
+                self.scope = TargetScope::Reject
+            }
+
+            match meta {
+                Meta::Path(p) => {
+                    if log_enabled!(log::Level::Warn) {
+                        warn!(
+                            "Encountered path while traversing target_os candidates: {}",
+                            p.into_token_stream()
+                        );
+                    }
+                }
+                Meta::List(meta_list) => {
+                    let nested_meta_list = meta_list
+                        .parse_args_with(Punctuated::<Meta, Token![,]>::parse_terminated)
+                        .inspect_err(|err| {
+                            error!("Failed to parse nested meta while traversing target_os candidates: {err}");
+                        })
+                        .ok()?;
+                    debug!("\texpanding with {} meta", nested_meta_list.len());
+                    self.meta.extend(nested_meta_list);
```

Comment:
A few questions here to make sure I'm following the recursive parsing correctly:

- Since we're using a queue of `Meta` nodes, it looks like we're effectively doing a BFS through all of the attributes, is that correct?
- Is the `self.scope` used to track when we're inside of a `not()` block, so all of the equating needs to be inverted?

If I'm understanding correctly, does this handle the case where the `not` is a portion of a list? E.g.:

```
#[cfg(all(not(feature = "my-feature"), target_os = "android"))]
```

It appears it would first recurse into the `all`, then hit the `not` and toggle the `scope` to `Reject`, and then hit the `target_os` (which is actually _outside_ of the `not`) but return it with a scope of `Reject`.

FWIW I think it's reasonable not to support the infinitely nested configs and instead handle a subset of them—as long as we clearly document what is and isn't supported with the flags. We definitely don't need to let the perfect be the enemy of the good here, but I also want to make sure we're not creating extra work for ourselves as maintainers going forward if we add a feature that only works in specific cases but don't indicate what those cases are.

Reply:
Updated scope handling and added this unit test.